exec/java-exec/src/test/java/org/apache/drill/test/OperatorTestBuilder.java

```diff
@@ -0,0 +1,300 @@
+    this.physicalOpUnitTestBase = physicalOpUnitTestBase;
+  }
+
+  @SuppressWarnings({"unchecked", "resource"})
+  public void go() throws Exception {
+    final List<RowSet> actualResults = new ArrayList<>();
+    CloseableRecordBatch testOperator = null;
+
+    try {
+      validate();
+      int expectedNumBatches = expectedNumBatchesOpt.orElse(expectedResults.size());
+      physicalOpUnitTestBase.mockOpContext(physicalOperator, initReservation, maxAllocation);
+
+      final BatchCreator<PhysicalOperator> opCreator = (BatchCreator<PhysicalOperator>) physicalOpUnitTestBase.opCreatorReg.getOperatorCreator(physicalOperator.getClass());
+      testOperator = opCreator.getBatch(physicalOpUnitTestBase.fragContext, physicalOperator, (List)upstreamBatches);
+
+      batchIterator: for (int batchIndex = 0;; batchIndex++) {
+        final RecordBatch.IterOutcome outcome = testOperator.next();
+
+        switch (outcome) {
+          case NONE:
+            if (!combineOutputBatches) {
+              Assert.assertEquals(expectedNumBatches, batchIndex);
+            }
+            // We are done iterating over batches. Now we need to compare them.
+            break batchIterator;
+          case OK_NEW_SCHEMA:
+            boolean skip = true;
+
+            try {
+              skip = testOperator.getContainer().getRecordCount() == 0;
+            } catch (IllegalStateException e) {
+              // We should skip this batch in this case. It means no data was included with the okay schema
+            } finally {
+              if (skip) {
+                batchIndex--;
+                break;
+              }
+            }
+          case OK:
+            if (!combineOutputBatches && batchIndex >= expectedNumBatches) {
+              testOperator.getContainer().clear();
+              Assert.fail("More batches received than expected.");
+            } else {
+              final boolean hasSelectionVector = testOperator.getSchema().getSelectionVectorMode().hasSelectionVector;
+              final VectorContainer container = testOperator.getContainer();
+
+              if (hasSelectionVector) {
+                actualResults.add(IndirectRowSet.fromSv2(testOperator.getContainer(), testOperator.getSelectionVector2()));
+              } else {
+                actualResults.add(DirectRowSet.fromContainer(container));
+              }
+              break;
+            }
+          default:
+            throw new UnsupportedOperationException("Can't handle this yet");
+        }
+      }
+
+      int actualTotalRows = actualResults.stream()
+        .map(rowSet -> rowSet.rowCount())
+        .reduce((Integer countA, Integer countB) -> countA + countB)
+        .orElse(0);
+
+      if (expectedResults.isEmpty()) {
+        Assert.assertEquals((int) expectedTotalRowsOpt.orElse(0), actualTotalRows);
+        // We are done, we don't have any expected result to compare
+        return;
+      }
+
+      if (combineOutputBatches) {
```

Comment:
what is the idea behind `combineOutputBatches` since looks like we are just copying data from actual received vector containers to a single container ?

Reply:
This is needed for operators like HashAgg. Internally HashAgg maintains multiple partitions, and on completion separate output batches are emitted for each partition. So if HashAgg is using 32 partitions internally, it will emit at least 32 batches even for a small test data set.

It is difficult to predict what records go to what partitions without going through the entire hashcode calculation for the test. Additionally I only care about the correctness of the final results, and not about what partition got what records. So combining all the results into a single batch for comparison simplifies testing.

Reply:
Thanks for explanation. But in case of combining batches I am not seeing actual comparison happening between actual and expected batch.

Reply:
Thanks for catching this. It looks like I deleted the compare statement during a refactor. This is a really stupid mistake, I'll add a test for OperatorTestBuilder to catch it.